pkg/aws/client/client_route53.go

```diff
@@ -91,83 +93,88 @@ func normalizeZoneId(zoneId string) string {
-	} else {
-		rrs = newResourceRecordSet(name, recordType, newResourceRecords(recordType, values), ttl)
+		// if there is any error, fallback to read/delete
 	}
-	if err := c.waitForRoute53RateLimiter(ctx); err != nil {
+	rrss, err := c.GetDNSRecordSets(ctx, zoneId, name, recordType)
+	if err != nil {
 		return err
 	}
-	_, err = c.Route53.ChangeResourceRecordSetsWithContext(ctx, newChangeResourceRecordSetsInput(zoneId, route53.ChangeActionDelete, rrs))
-	if isValuesDoNotMatchError(err) && len(values) > 0 && ttl > 0 {
-		// The actual values / ttl are different from the given values / ttl
-		// Get the resource recordset from the zone and try again
-		rrs, err = c.GetDNSRecordSet(ctx, zoneId, name, getRecordType(recordType, values[0]))
-		if err != nil {
-			return err
-		}
-		if rrs == nil {
-			return nil
-		}
-		if err := c.waitForRoute53RateLimiter(ctx); err != nil {
-			return err
-		}
-		_, err = c.Route53.ChangeResourceRecordSetsWithContext(ctx, newChangeResourceRecordSetsInput(zoneId, route53.ChangeActionDelete, rrs))
+	if len(rrss) == 0 {
+		return nil
 	}
+	if err := c.waitForRoute53RateLimiter(ctx); err != nil {
+		return err
+	}
+	_, err = c.Route53.ChangeResourceRecordSetsWithContext(ctx, newChangeResourceRecordSetsInput(zoneId, route53.ChangeActionDelete, rrss))
 	return ignoreResourceRecordSetNotFound(err)
 }
 
-// GetDNSRecordSet returns the DNS recordset in the DNS hosted zone with the given zone ID, and with the given name and type.
-func (c *Client) GetDNSRecordSet(ctx context.Context, zoneId, name, recordType string) (*route53.ResourceRecordSet, error) {
+// GetDNSRecordSets returns the DNS recordset(s) in the DNS hosted zone with the given zone ID, and with the given name and type.
+// For record type CNAME there may be multiple DNS recordsets if mapped to alias targets A or AAAA recordsets.
+func (c *Client) GetDNSRecordSets(ctx context.Context, zoneId, name, recordType string) ([]*route53.ResourceRecordSet, error) {
 	if err := c.waitForRoute53RateLimiter(ctx); err != nil {
 		return nil, err
 	}
-	out, err := c.Route53.ListResourceRecordSetsWithContext(ctx, &route53.ListResourceRecordSetsInput{
+	input := &route53.ListResourceRecordSetsInput{
 		HostedZoneId:    aws.String(zoneId),
 		MaxItems:        aws.String("1"),
 		StartRecordName: aws.String(name),
 		StartRecordType: aws.String(recordType),
-	})
+	}
+	if recordType == route53.RRTypeCname {
+		input.MaxItems = aws.String("5") // potential CNAME, AliasTarget A and AliasTarget AAAA
+		input.StartRecordType = nil
+	}
+	out, err := c.Route53.ListResourceRecordSetsWithContext(ctx, input)
 	if ignoreResourceRecordSetNotFound(err) != nil {
 		return nil, err
 	}
 	if out == nil || len(out.ResourceRecordSets) == 0 { // no records in zone
 		return nil, nil
 	}
-	if rrs := out.ResourceRecordSets[0]; normalizeName(aws.StringValue(rrs.Name)) == name && aws.StringValue(rrs.Type) == recordType {
-		return out.ResourceRecordSets[0], nil
+	var toDelete []*route53.ResourceRecordSet
```

Comment:
Why is this variable called `toDelete` when it has nothing to do with deletion in this getter function? The function result may be used for deletion, but here I find the name rather confusing.

Reply:
Good point. Renamed it to `recordSets` to avoid imposing purpose of method call.